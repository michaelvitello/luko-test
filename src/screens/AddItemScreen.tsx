import React, { useState } from 'react'
import { useForm, Controller } from 'react-hook-form'
import {
    StyleSheet,
    View,
    KeyboardAvoidingView,
    TextInput,
    Pressable,
    Platform,
    Text,
    Image,
} from 'react-native'
import Button from '../components/Button'
import { RootTabScreenProps } from '../navigation/types'
import { ImagePicker } from '../sdk/ImagePicker'
import { colors } from '../theme/colors'
import { fonts } from '../theme/fonts'

export default function AddItemScreen({ navigation }: RootTabScreenProps<'AddItemScreen'>) {
    /* Initialize Form for item addition */
    const {
        control,
        handleSubmit,
        formState: { errors, isValid },
    } = useForm({
        defaultValues: {
            name: '',
            value: '',
            description: '',
        },
    })

    /* Image state for item picture */
    const [image, setImage] = useState(null)

    /* Form submission - TODO: Write in DB to update list */
    const onSubmit = data => {
        // console.log(data) -> Logged data for debug
        navigation.goBack()
    }

    /* Image picker - TODO: Same with image capture (camera) */
    const imagePick = async () => {
        const result = await ImagePicker.pickImage()
        setImage(result.assets[0].uri)
    }

    /* Use enable, not disable for boolean JS ordering - we need true && true not false && true, which is false */
    const enableAddButton = isValid && image

    return (
        /* TODO - Fix Keyboard not overlapping view */
        /* Try KeyboardAwareScrollView ? https://github.com/APSL/react-native-keyboard-aware-scroll-view */
        <KeyboardAvoidingView
            behavior={Platform.OS === 'ios' ? 'padding' : 'height'}
            contentContainerStyle={{ flexGrow: 1 }} // Not sure that works...
            style={styles.container}
        >
            <View style={styles.buttonsContainer}>
                <Button onPress={() => navigation.goBack()} title='Cancel' />
                <Button disabled={!enableAddButton} onPress={handleSubmit(onSubmit)} title='Add' />
            </View>
            <View style={styles.imagePicker}>
                <Pressable
                    onPress={imagePick}
                    pressRetentionOffset={10}
                    style={({ pressed }) => [{ opacity: pressed ? 0.5 : 1 }, styles.imageButton]}
                >
                    {image ? (
                        <Image source={{ uri: image }} style={styles.image} />
                    ) : (
                        <Text numberOfLines={1} style={styles.title}>
                            {'Add Photo'}
                        </Text>
                    )}
                </Pressable>
            </View>
            <Controller
                control={control}
                name='name'
                render={({ field: { onChange, onBlur, value } }) => (
                    <TextInput
                        onBlur={onBlur}
                        onChangeText={onChange}
                        placeholder=' Name'
                        style={styles.input}
                        value={value}
                    />
                )}
                rules={{
                    required: true,
                }}
            />
            {errors.name && <Text>This is required.</Text>}
            <Controller
                control={control}
                name='value'
                render={({ field: { onChange, onBlur, value } }) => (
                    <TextInput
                        keyboardType='numeric'
                        onBlur={onBlur}
                        onChangeText={onChange}
                        placeholder=' 0 â‚¬'
                        style={styles.input}
                        value={value}
                    />
                )}
                rules={{
                    max: 40000,
                    min: 0,
                    required: true,
                    maxLength: 1000,
                }}
            />
            <Controller
                control={control}
                name='description'
                render={({ field: { onChange, onBlur, value } }) => (
                    <TextInput
                        multiline={true}
                        onBlur={onBlur}
                        onChangeText={onChange}
                        placeholder=' Description (optional)'
                        style={styles.descriptionInput}
                        value={value}
                    />
                )}
            />
        </KeyboardAvoidingView>
    )
}

const styles = StyleSheet.create({
    container: {
        flex: 1,
        paddingHorizontal: 20,
        backgroundColor: colors.background,
        paddingTop: 10,
    },
    buttonsContainer: {
        width: '100%',
        marginTop: 16,
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
        backgroundColor: 'transparent',
    },
    formContainer: {
        width: '100%',
        height: 400,
        borderColor: 'yellow',
        borderWidth: 2,
    },
    input: {
        height: 40,
        width: '100%',
        borderWidth: 1,
        borderColor: colors.mainGrey,
        borderRadius: 5,
        marginBottom: 50,
        fontFamily: fonts.bold,
        fontSize: 12,
        paddingHorizontal: 10,
        color: 'black',
    },
    descriptionInput: {
        height: 120,
        width: '100%',
        borderWidth: 1,
        borderColor: colors.mainGrey,
        borderRadius: 5,
        fontFamily: fonts.bold,
        fontSize: 12,
        paddingHorizontal: 10,
        color: 'black',
    },
    imagePicker: {
        marginTop: 50,
        marginBottom: 50,
        width: '100%',
        height: '20%',
        alignItems: 'center',
        justifyContent: 'center',
        overflow: 'hidden',
    },
    image: {
        width: '100%',
        height: '100%',
    },
    imageButton: {
        width: 100,
        height: 100,
        borderRadius: 50,
        borderWidth: 5,
        borderColor: colors.mainBlue,
        alignItems: 'center',
        justifyContent: 'center',
        overflow: 'hidden',
    },
    title: {
        fontFamily: fonts.bold,
        fontSize: 12,
        color: colors.mainGrey,
    },
})
