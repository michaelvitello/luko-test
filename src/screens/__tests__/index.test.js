import React from 'react'
import renderer from 'react-test-renderer'
import AddItemScreen from '../AddItemScreen'

test('renders AddItemScreen correctly', () => {
    const tree = renderer.create(<AddItemScreen />).toJSON()
    expect(tree).toMatchSnapshot()
})


