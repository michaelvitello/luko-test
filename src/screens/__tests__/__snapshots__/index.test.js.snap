// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#F4F3EF",
        "flex": 1,
        "paddingHorizontal": 20,
        "paddingTop": 10,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "transparent",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginTop": 16,
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={20}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "opacity": 1,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#2D50E6",
            "fontSize": 17,
          }
        }
      >
        Cancel
      </Text>
    </View>
    <View
      accessibilityState={
        Object {
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      hitSlop={20}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "opacity": 1,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "color": "#C0BEB8",
            "fontSize": 17,
          }
        }
      >
        Add
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "height": "20%",
        "justifyContent": "center",
        "marginBottom": 50,
        "marginTop": 50,
        "overflow": "hidden",
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Array [
          Object {
            "opacity": 1,
          },
          Object {
            "alignItems": "center",
            "borderColor": "#2D50E6",
            "borderRadius": 60,
            "borderWidth": 2,
            "height": 120,
            "justifyContent": "center",
            "overflow": "hidden",
            "width": 120,
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        style={
          Object {
            "color": "#C0BEB8",
            "fontFamily": "Circular-Bold",
            "fontSize": 12,
          }
        }
      >
        Add Photo
      </Text>
    </View>
    <Text
      numberOfLines={1}
      style={
        Array [
          Object {
            "fontFamily": "Circular-Bold",
            "fontSize": 10,
            "paddingTop": 10,
          },
          Object {
            "color": "red",
          },
        ]
      }
    >
      Mandatory
    </Text>
  </View>
  <Text
    numberOfLines={1}
    style={
      Object {
        "color": "black",
        "fontFamily": "Circular-Bold",
        "fontSize": 12,
        "paddingBottom": 10,
      }
    }
  >
    Name
  </Text>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder=" Bracelet"
    style={
      Array [
        Object {
          "borderRadius": 5,
          "borderWidth": 1,
          "color": "black",
          "fontFamily": "Circular-Bold",
          "fontSize": 12,
          "height": 40,
          "marginBottom": 20,
          "paddingHorizontal": 10,
          "width": "100%",
        },
        Object {
          "borderColor": "#C0BEB8",
        },
      ]
    }
    value=""
  />
  <Text
    numberOfLines={1}
    style={
      Object {
        "color": "black",
        "fontFamily": "Circular-Bold",
        "fontSize": 12,
        "paddingBottom": 10,
      }
    }
  >
    Value
  </Text>
  <TextInput
    keyboardType="numeric"
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder=" 0 â‚¬"
    style={
      Array [
        Object {
          "borderRadius": 5,
          "borderWidth": 1,
          "color": "black",
          "fontFamily": "Circular-Bold",
          "fontSize": 12,
          "height": 40,
          "marginBottom": 20,
          "paddingHorizontal": 10,
          "width": "100%",
        },
        Object {
          "borderColor": "#C0BEB8",
        },
      ]
    }
    value=""
  />
  <Text
    numberOfLines={1}
    style={
      Object {
        "color": "black",
        "fontFamily": "Circular-Bold",
        "fontSize": 12,
        "paddingBottom": 10,
      }
    }
  >
    Description
  </Text>
  <TextInput
    multiline={true}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder=" Optional"
    style={
      Object {
        "borderColor": "#C0BEB8",
        "borderRadius": 5,
        "borderWidth": 1,
        "color": "black",
        "fontFamily": "Circular-Bold",
        "fontSize": 12,
        "height": 120,
        "paddingHorizontal": 10,
        "width": "100%",
      }
    }
    value=""
  />
</View>
`;
